hw5_to_do_list/oop/js/components.js

```diff
@@ -0,0 +1,361 @@
+        this.state = {
+            items: this.hasSavedState() ? this.retrieveFromLocalStorage() : []
+        }
+        this.title = new Header({ text: props.title, level: 2 })
+    }
+
+    updateDOM() {
+        this.saveToLocalStorage()
+        this.children = [
+            this.title,
+            ...this.state.items
+        ]
+    }
+
+    removeItem = item => {
+        let items = [...this.state.items]
+        const idx = items.findIndex(i => { i.rawItem == item.rawItem })
+        items.splice(idx, 1)
+        this.setState({
+            items: items
+        })
+    }
+
+    addItem = item => {
+        const rawItem = item instanceof ListItem ? item.rawItem : item
+        this.setState({
+            items: this.state.items.concat(this.wrapRawItem(rawItem))
+        })
+    }
+
+    wrapRawItem = rawItem => new ListItem({
+        rawItem: rawItem,
+        removeItem: this.removeItem,
+        cssClass: this.props.itemCssClass,
+        onItemCheckboxClick: this.props.onItemCheckboxClick,
+    })
+
+
+    displayBySubstring = substring => {
+        this.state.items.forEach(
+            item => {
+                item.setState({
+                    isDisplayed: item.name.text.toLowerCase().includes(substring)
+                })
+            }
+        )
+    }
+
+    saveToLocalStorage() {
+        const rawItems = this.state.items.map(item => item.rawItem)
+        localStorage.setItem(this.props.storageKey, JSON.stringify(rawItems))
+    }
+
+    retrieveFromLocalStorage() {
+        const rawItems = JSON.parse(localStorage.getItem(this.props.storageKey))
+        return rawItems.map(rawItem => this.wrapRawItem(rawItem))
+    }
+}
+
+
+
+class ListItem extends Component {
+    constructor(props) {
+        super(props)
+        this.state = {
+            isDisplayed: true
+        }
+        this.rawItem = props.rawItem
+        this.rawItem.isCompleted = this.rawItem.isCompleted || false
+        this.checkbox = new Checkbox({
+            onChange: () => props.onItemCheckboxClick(this),
+            cssClass: 'checkbox__input',
+            isChecked: this.rawItem.isCompleted,
+        })
+        this.name = new Header({
+            cssClass: 'checkbox__task-title',
+            text: this.rawItem.name,
+            level: 3
+        })
+        this.label = new Component({
+            cssClass: `label label-${this.rawItem.label}`,
+            children: [this.rawItem.label]
+        })
+        this.date = new Component({
+            cssClass: 'date',
+            children: [this.formatDate(this.rawItem.date)]
+        })
+        this.removeButton = new Button({
+            onClick: () => props.removeItem(this),
+            cssClass: 'checkbox__delete-btn'
+        })
+    }
+
+    updateDOM() {
+        if (this.state.isDisplayed) {
+            this.display()
+        } else {
+            this.hide()
+        }
+
+        this.children = [
+            new Component({
```

Comment:
You don't need to create Component class instance for each `div`  you need on the page.
Sometimes it is enough to call document.createElement(...).append(...)